feat(File Upload): Add retry and upload a new file workflow under feature flag

dat-gcio opened the pull request:
## Description
Implement new workflow to reduce the number of actions veterans have to take in order to retry, or upload a new file when an error happens.

## Testing done
- [ ] unit
- [ ] manual

## Screenshots

File type not allowed:
![image](https://user-images.githubusercontent.com/83595736/123308427-a5468080-d4f1-11eb-99b2-65f7ee5f3afe.png)

File too large:
![image](https://user-images.githubusercontent.com/83595736/123308475-b8595080-d4f1-11eb-9cf6-4419bb644c82.png)

File too small: 
![image](https://user-images.githubusercontent.com/83595736/123308526-c7400300-d4f1-11eb-800b-98af4cfe09f9.png)

Network error:
![image](https://user-images.githubusercontent.com/83595736/123308577-d58e1f00-d4f1-11eb-95ee-043246e87f9b.png)

Don't allow additional upload if error present:
Before:
![image](https://user-images.githubusercontent.com/83595736/123309073-7381e980-d4f2-11eb-8138-668bf980c946.png)

Now: 
![image](https://user-images.githubusercontent.com/83595736/123308734-079f8100-d4f2-11eb-8553-01af02821b59.png)


## Acceptance criteria
- [ ] Update connection error:
We're sorry. We had a connection problem. Please try again.
[Try again]  [Cancel]

- [ ] Update File too large error:
We couldn't upload your file because it's too big. Please make sure the file is 10MB or less and try again.
[Upload a new file] [Cancel]

- [ ] Update File too small error:
We couldn't upload your file because it's too small. Please make sure the file is 1B or more and try again.
[Upload a new file] [Cancel]

- [ ] Update Wrong file type error
We couldn't upload your file because we can't accept this type of file. Please make sure the file is a .pdf, .jpeg, jpg, or .png file and try again.
[Upload a new file] [Cancel]

- [ ] Forms with multiple file upload option should remove "Add another button" on error
- [ ] Try again should resend the file to the backend
- [ ] Upload another file should delete the file, and prompt for file to upload






## Definition of done
- [ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/26419)

dat-gcio committed:
feat(File Upload): Add retry and upload a new file workflow under feature flag

dat-gcio committed:
update error messages to match copy

dat-gcio committed:
fix up unit test

dat-gcio commented:
Not yet ready for review. Accidentally clicked.

dat-gcio committed:
make upload another file button primary

dat-gcio committed:
update test to use character code to distinguish

dat-gcio committed:
guard new feature under toggle

dat-gcio committed:
consolidate buttons and improve error accessibility

dat-gcio committed:
conslidate button and add more flags

dat-gcio commented:
> Not yet ready for review. Accidentally clicked. Meant to open review instance link.
PR ready for review

dat-gcio committed:
reduce cognitive complexity to satisfy check

dat-gcio committed:
fix conditional with allowUploadWhenError

bkjohnson commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -237,6 +265,17 @@ class FileField extends React.Component {
+            [file.errorMessage].filter(error => error);
+
+          return errors.length > 0;
+        }));
```
Why are uploads being allowed if there are errors?

dat-gcio replied:
It currently allows it, so I wrapped this 'enhancement' under the feature flag.

![image](https://user-images.githubusercontent.com/83595736/123857944-f2ab5f00-d8f0-11eb-8447-5c63f728e348.png)

bkjohnson reviewed (commented):
This is a first round - I'll come back again and leave some more comments/questions.

dat-gcio commented on src/platform/forms-system/src/js/actions.js:
```diff
@@ -335,12 +349,19 @@ export function uploadFile(
       if (password) {
-        onChange({ name: file.name, errorMessage, password: file.password });
+        onChange({
+          file, // return file object to allow resubmit
+          name: file.name,
+          errorMessage,
+          password: file.password,
+        });
       } else {
-        onChange({ name: file.name, errorMessage });
+        onChange({ file, name: file.name, errorMessage }); // return file object to allow resubmit
```
The 'Please delete this file' and try again verbiage was added in the last iteration. It is to provide users with direct directions. It is helpful keep it, when the new button/workflow is not enabled.

dat-gcio commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -359,8 +412,8 @@ class FileField extends React.Component {
                     hasErrors && (
-                      <span className="usa-input-error-message">
-                        {errors[0]}
+                      <span className="usa-input-error-message" role="alert">
+                        <span className="sr-only">Error</span> {errors[0]}
```
Ah, that makes sense. I'll keep it in mind for next time.

dat-gcio replied:
You are right, it is out of scope. Opted to remove it, and do it in another PR linking to that ticket.

dat-gcio committed:
add comment and remove unneeded flag guard

dat-gcio commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -184,7 +199,19 @@ class FileField extends React.Component {
     }
-    this.focusAddAnotherButton();
+
+    if (focusAddButton) {
+      this.focusAddAnotherButton();
+    }
```
Good call. Added.

dat-gcio commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -109,11 +118,16 @@ class FileField extends React.Component {
+              name: currentFile.name,
+              isEncrypted: true,
+            };
+          }
```
Ended up removing this change, since it doesn't need guarding.  The only time we need the 'file' object is on resending and the action.js function takes care of sending the file back anyways.

bkjohnson commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+        />,
+      );
+
+      const cancelButton = tree.find('button.usa-button-secondary');
+      expect(cancelButton.text()).to.equal('Delete file');
```
Naming.

dat-gcio replied:
Updated name, and added comment.

bkjohnson commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+        />,
+      );
+
+      const uploadButton = tree.find('#field_add_label');
+      expect(uploadButton.exists()).to.be.true;
```
Same comment here about the selector - I'm not sure what it's referring to.

dat-gcio replied:
Added a comment to help discern that the id is interpolated.

bkjohnson commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+        />,
+      );
+
+      const cancelButton = tree.find('button.usa-button-secondary');
+      expect(cancelButton.text()).to.equal('Delete file');
```
Naming.

dat-gcio replied:
Updated and added comment

bkjohnson commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+        />,
+      );
+
+      const cancelButton = tree.find('button.usa-button-secondary');
+      expect(cancelButton.text()).to.equal('Delete file');
```
Naming.

bkjohnson commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+        />,
+      );
+
+      const retryButton = tree.find('button.usa-button-primary');
+      expect(retryButton.exists()).to.be.false;
```
It's not obvious from this that the "retry" button is a primary button - can you make that more clear?

dat-gcio replied:
Added a comment to help clarify. Let me know if you meant something else.

bkjohnson reviewed (changes requested):
In the tests for `FileField`, I think it would be good to move some of the `const`s that don't change between tests (`idSchema`, `schema`) up to the `describe` level rather than redefine them in each `it` block.

dat-gcio committed:
refactor filefield unit test

dat-gcio committed:
add more comments to unit test

dat-gcio commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -388,42 +441,45 @@ class FileField extends React.Component {
-                  aria-label={`${buttonText} ${titleString}`}
+        {// Don't render an upload button on review & submit page while in
+        // review mode
+        showButtons && (
+          <>
+            {(maxItems === null || files.length < maxItems) &&
```
I needed the actual file input in the DOM even when the main file 'upload' button is not present. It allows the 'Retry' button to work properly.

dat-gcio commented on src/platform/forms-system/test/js/actions.unit.spec.js:
```diff
@@ -286,6 +286,7 @@ describe('Schemaform actions:', () => {
   describe('uploadFile', () => {
```
Updated with your suggestion

dat-gcio commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+      );
+
+      const uploadButton = tree.find('#field_add_label');
+      expect(uploadButton.exists()).to.be.false;
```
I am targeting the label that wraps that span. I renamed the idSchema.$id prop in the test to help identify it. Let me if that is not what you meant.

dat-gcio commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+      );
+
+      const cancelButton = tree.find('button.usa-button-primary');
+      expect(cancelButton.text()).to.equal('Upload a new file');
```
Updated name, and added comment.

dat-gcio commented on src/platform/forms-system/test/js/fields/FileField.unit.spec.jsx:
```diff
@@ -919,4 +920,426 @@ describe('Schemaform <FileField>', () => {
+      );
+
+      const cancelButton = tree.find('button.usa-button-primary');
+      expect(cancelButton.text()).to.equal('Try again');
```
Updated name, and added comment.

dat-gcio commented:
> In the tests for `FileField`, I think it would be good to move some of the `const`s that don't change between tests (`idSchema`, `schema`) up to the `describe` level rather than redefine them in each `it` block.

Thanks for the suggestion. Moved the reused `const`s up to the `describe` level. Also updated the test description to be more descriptive and added comments above the button queries to describe the target. I was too caught up in handling all the cases of the feature flag, and should have done a better job with the test writing itself. Sorry you had to go through all of that. 
![image](https://user-images.githubusercontent.com/83595736/124143847-74b69780-da59-11eb-93db-9269f98b9fdd.png)

timwright12 commented:
Don't have a lot of context into these updates, so I'll defer to @bkjohnson for PR approval. Let me know if you do need me though.

bkjohnson commented on src/platform/forms-system/src/js/fields/FileField.jsx:
```diff
@@ -237,6 +255,17 @@ class FileField extends React.Component {
         ? uiSchema['ui:title']
         : schema.title;
 
+    const allowUploadWhenError =
+      !enableShortWorkflow ||
+      (enableShortWorkflow &&
+        !files.some((file, index) => {
+          const errors =
+            _.get([index, '__errors'], errorSchema) ||
+            [file.errorMessage].filter(error => error);
+
+          return errors.length > 0;
+        }));
```
Does the condition after the `||` mean that this would allow uploads when there _aren't_ errors? I see the `!` has been added to the beginning of `files`.

dat-gcio replied:
Yup. If the flag is on, only allow additional uploads when there are no errors present.

bkjohnson replied:
Does the `WhenError` part of the name still make sense?

dat-gcio replied:
Because the feature flag is mixed in there, it does. The variable will need a rename when the flag is removed. With the flag off (`!enableShortWorkflow`) we `allowUploadWhenError`. When the flag is on, we don't `allowUploadWhenError`.

bkjohnson replied:
I think this should be renamed to reflect what the value represents regardless of whether or not the feature flag is flipped (`allowUpload` possibly), and a comment added explaining how the feature flag changes the criteria for allowing an upload.

This is defined on line 258 and not used until line 450. If a future person is reading this code and sees `allowUploadWhenError && ...`, there's a good chance that they may not be thinking "but invert this if a specific feature flag is on", because the definition won't be visible. Even if the feature flag is removed and becomes the default a week from now, I'd rather avoid some potential awkwardness with the naming.

dat-gcio replied:
Ok, finally understood what you were telling me. Name changed and comment added. And no worries on not making staging review. I don't think anyone on my team knew we had to get the artifact in staging a week before the staging review date. We all thought it was by the date of the staging review. 

And I really appreciate your thoroughness, especially since I am very new to va.gov.

dat-gcio commented:
Hey @bkjohnson, do you think this can be approved by 2PM EST today? Need to have it in staging for staging review.

dat-gcio commented:
> Hey @bkjohnson, do you think this can be approved by 2PM EST today? Need to have it in staging for staging review.

Staging review has been pushed back a week, so we have until Thursday to get this in.

bkjohnson reviewed (commented):
Hey, sorry this missed staging review. I don't mean to be too nitpicky, but the forms library already has some problems with clarity/complexity, and I'm trying to make sure that others who read the code will not have too much difficulty in understanding what's going on.

The nature of the forms library also means that making small PRs can be hard or impossible, and longer PRs simply take more time to review. If a PR _is_ long, leaving your own review can help speed things up when another codeowner steps in to review it.

Thanks for your patience!

dat-gcio committed:
fix variable name and add comment to help explain feature gate

bkjohnson reviewed (approved):
:tada:
